src/library/scala/collection/Map.scala

```diff
@@ -179,9 +179,10 @@ trait MapOps[K, +V, +CC[_, _] <: IterableOps[_, AnyConstr, _], +C]
 
   override /*PartialFunction*/ def applyOrElse[K1 <: K, V1 >: V](x: K1, default: K1 => V1): V1 = getOrElse(x, default(x))
 
-  /** Collects all keys of this map in a set.
-    * @return  a set containing all keys of this map.
-    */
+  /** A set view of the keys contained by this map.
+   *
+   *  @return  a set view of the keys contained by this map
+   */
   def keySet: Set[K] = new KeySet
```

Comment:
Questions:
1. will remove the key from a keyset affect the underling map?
2. can user add new key to the returning keySet?
3. does the implementation cache a keyset instance ?

Reply:
I think the semantic question is important: do we guarantee that `keySet` returns a view? Because that's not reflected in the return type.

It might be better to say that it *may* return a view which *may* hold a reference to the original collection, and *may* reflect changes to the underlying Map.

Reply:
it is a `collection.Set` so answers are provided by the type, unlike Java. The other active question is the ticket about treemap.range.clear() clearing the map in scala but only the subrange in java.

By uniform access, it's an implementation detail whether the same instance is returned always.

IIUC, the collections (by Mr Reginald Straw, the straw-man) intend for subclasses to return a subclass of `KeySet`; it is appropriate that the class is not exposed in the API. Arguably, it's appropriate for the doc to announce the most severe restriction (it's a view because that is efficient) which may be relaxed by any given subclass (not a view, maybe it's easier to return a new set of one element than a view object that must walk an extra pointer).

Reply:
> It might be better to say that it may return a view which may hold a reference to the original collection, and may reflect changes to the underlying Map.

Isn't it weird to say that `A` _may_ return `b` ? I think sometimes we would like to be sure not to have a reference to the original collection

Reply:
Some implementations return a view, some overrides return a new independent collection. I don't think this is good design, but it's what we have and we probably cannot change it in a minor release. We can add a snippet to the documentation showing how to turn the result into a strict copy.